numba/np/polynomial/polynomial_core.py

```diff
@@ -0,0 +1,218 @@
+    two = i64(2)
+
+    s1 = builder.extract_value(domain_helper.shape, 0)
+    s2 = builder.extract_value(window_helper.shape, 0)
+    pred = builder.or_(
+        builder.icmp_signed('!=', s1, two),
+        builder.icmp_signed('!=', s2, two))
+
+    with cgutils.if_unlikely(builder, pred):
+        context.call_conv.return_user_exc(
+            builder, ValueError,
+            ("Domain and Window must be of length 2.",))
+
+    polynomial.coef = coef_cast
+    polynomial.domain = domain_helper._getvalue()
+    polynomial.window = window_helper._getvalue()
+
+    return polynomial._getvalue()
+
+
+@unbox(types.PolynomialType)
+def unbox_polynomial(typ, obj, c):
+    """
+    Convert a Polynomial object to a native polynomial structure.
+    """
+    is_error_ptr = cgutils.alloca_once_value(c.builder, cgutils.false_bit)
+    polynomial = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+    with ExitStack() as stack:
+        coef_obj = c.pyapi.object_getattr_string(obj, "coef")
+        with cgutils.early_exit_if_null(c.builder, stack, coef_obj):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+        coef_native = c.unbox(typ.coef, coef_obj)
+        c.pyapi.decref(coef_obj)
+        with cgutils.early_exit_if(c.builder, stack, coef_native.is_error):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+
+        domain_obj = c.pyapi.object_getattr_string(obj, "domain")
+        with cgutils.early_exit_if_null(c.builder, stack, domain_obj):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+        domain_native = c.unbox(typ.domain, domain_obj)
+        c.pyapi.decref(domain_obj)
+        with cgutils.early_exit_if(c.builder, stack, domain_native.is_error):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+
+        window_obj = c.pyapi.object_getattr_string(obj, "window")
+        with cgutils.early_exit_if_null(c.builder, stack, window_obj):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+        window_native = c.unbox(typ.window, window_obj)
+        c.pyapi.decref(window_obj)
+        with cgutils.early_exit_if(c.builder, stack, window_native.is_error):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+
+        polynomial.coef = coef_native.value
+        polynomial.domain = domain_native.value
+        polynomial.window = window_native.value
+
+    return NativeValue(polynomial._getvalue(),
+                       is_error=c.builder.load(is_error_ptr))
+
+
+@box(types.PolynomialType)
+def box_polynomial(typ, val, c):
+    """
+    Convert a native polynomial structure to a Polynomial object.
+    """
+    ret_ptr = cgutils.alloca_once(c.builder, c.pyapi.pyobj)
+    fail_obj = c.pyapi.get_null_object()
+
+    with ExitStack() as stack:
+        polynomial = cgutils.create_struct_proxy(typ)(c.context, c.builder,
+                                                      value=val)
+        coef_obj = c.box(types.Array(types.double, 1, 'C'), polynomial.coef)
```

Comment:
`coeficient` array dtype is always a double, right?
```suggestion
        coef_obj = c.box(typ.coef, polynomial.coef)
```